### ISSUE TITLE
rfc: use codeowners to automatically request reviews

### ISSUE THREAD

**Issue by SethFalco** [tldr-pages/tldr#9340]
Currently, we're frequently requesting reviews to maintainers, seemingly randomly, to get pull requests looked at sooner.

I'd like to propose that we reduce this practice, in favor of a more opt-in approach where we use a CODEOWNERS file to automatically request reviews on pages based on the maintainers interests.

The idea is that, if someone is primarily interested in reviewing pages for a certain language, operating system, or maybe the GitHub health files, they can opt-in for that.

This has a few benefits:
1. The active reviewers will no longer have to request reviews from other maintainers as frequently.
2. Less active reviewers (such as myself) may be more inclined to respond to review requests when it's something we're actively interested in.

Docs:
https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners

The file would look something like this:

```txt
# Default
* {whoever is interested in anything not otherwise matched by anything below}

# Languages
/pages.de/ @tldr-pages/german 
/pages.es/ @tldr-pages/spanish 

# Language / Operating System
linux/ {whoever is interested in all linux pages regardless of language}

/pages/**/linux/ {whoever is interested in english pages for linux tools}
/pages/**/common/ {whoever is interested in english pages for common tools} 

# Pages
/pages/**/register_new_matrix_user.md @SethFalco 

# Health Files
/*.md @SethFalco 
/contributing-guides/*.md @SethFalco 

# Dependencies
/*.json {whoever is interested in linters/ci}
/requirements.txt {whoever is interested in linters/ci}
```
> Note that each line in the CODEOWNERS file can have more than one team/user in it.

We can always play with this whenever we want, or remove names when someone wants to opt out of being requested for review. We can't allow the file to exceed 3 MB, so we can't have too much fun with page-specific code owners, but overall, I don't expect that to be a problem.

The only caveat I see is that with how this is handled, if something further down matches a code owner, none of the matches above will count. When defining this file, I recommend starting with the most abstract file groups and working your way to be more specific.

Note above how I handle the `Language / Operating System` section in the example. I specify the OS first, then only after, language-specific versions for the OS. If it was written the other way around, then `/pages/**/linux/` would be redundant as it'd _always_ match `linux/` afterwards which would steal the review.

This also means, however, that the global reviewers at the top will not be requested to review if something further down matches. (Unless that global reviewer is also in the list of the last matched glob.)

What are our overall thoughts on this?

While it's not as clean as I'd like, I think this is a solid approach as it gives maintainers more control over when they're requested to review tasks.

None of this means that we can't request reviews from other maintainers anymore. Just that hopefully we can automate some of it away so maintainers get the things more interesting to them, and we can fallback to manual requests if things aren't getting reviewed.

**Comment by kbdharun** [comment]
Excellent proposal, I have a doubt would this prevent the merging of PRs for reviewers who aren't code owners?

**Comment by SethFalco** [comment]
> Repository owners can add branch protection rules to ensure that changed code is reviewed by the owners of the changed files.
> 
> — https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners#codeowners-and-branch-protection

> Optionally, you can choose to require reviews from code owners. If you do, any pull request that affects code with a code owner must be approved by that code owner before the pull request can be merged into the protected branch.
> 
> — https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/about-protected-branches

This gives me the impression that by default, code owners do are not blocking. They'll be requested by default, but do not _have to_ provide a review.

**Comment by kbdharun** [comment]
> This gives me the impression that by default, code owners are not blocking. They'll be requested by default, but do not have to provide a review.

Nice, Then I think we can use it I guess as by default it will not prevent merge.

**Comment by waldyrious** [comment]
> The only caveat I see is that with how this is handled, if something further down matches a code owner, none of the matches above will count. When defining this file, I recommend starting with the most abstract file groups and working your way to be more specific.
> 
> Note above how I handle the `Language / Operating System` section in the example. I specify the OS first, then only after, language-specific versions for the OS. If it was written the other way around, then `/pages/**/linux/` would be redundant as it'd _always_ match `linux/` afterwards which would steal the review.
> 
> This also means, however, that the global reviewers at the top will not be requested to review if something further down matches. (Unless that global reviewer is also in the list of the last matched glob.)

Weird, that seems a little counterintuitive. I would expect the stewardship (to step away from the "ownership" lingo) specifications would be cumulative. Are you aware of the rationale for this behavior, or any discussions about changing it? Perhaps it stems precisely from an understanding of the people as "owners" of the code, in which case I can see why people would prefer more specific rules to override more generic ones. But it seems like something that should be configurable.

**Comment by SethFalco** [comment]
> specifications would be cumulative

You're right imo, at least I'd expect there to be an option to choose between replacing vs accumlating reviewers.

This is my first time exploring this feature and reading the docs, but unfortunately, I couldn't find anything in the docs to make it accumulate reviewers.

A (messy) workaround would be to make a team called @tldr-pages/reviewers which we put on every match. (Note that we can list multiple team/users after a matcher.)

> Are you aware of the rationale for this behavior, or any discussions about changing it?

In the announcement and docs, I see no explicit explanation of why they implemented it this way. If I were to hazard a **_guess_**, it may be because the syntax is based on `.gitattributes`:

> This format may be familiar if you’ve used [gitattributes](https://git-scm.com/docs/gitattributes).
> 
> — https://github.blog/2017-07-06-introducing-code-owners/

> When more than one pattern matches the path, a later line overrides an earlier line.
> 
> — https://git-scm.com/docs/gitattributes

**Comment by kbdharun** [comment]
Same here this is my first time exploring it too. I had only seen this used in archinstall from what I remember.

**Comment by EmilySeville7cfg** [comment]
I love such idea. I wanna review Linux pages and their translations more often then the other ones. :)
